Add the ability to use external definitions for monostate and bad variant access

MostafaNanticock opened the pull request:
# Description

We are using variant-lite in a project alongside `tl::optional` and other header-only C++11+ compatibility libraries. To provide a unified interface for downstream users, we expose them under a `std_compat` namespace:
```cpp
namespace std_compat
{
    template <class T>
    using optional = tl::optional<T>;

    using tl::bad_optional_access;
    using tl::in_place_t;
    using tl::in_place;
    using tl::nullopt;
    using tl::nullopt_t;
    using tl::make_optional;
    using tl::swap;
}
```
```cpp
namespace std_compat
{
    template <class... Types>
    using variant = nonstd::variant<Types...>;

    using nonstd::bad_variant_access;
    using nonstd::get;
    using nonstd::get_if;
    using nonstd::holds_alternative;
    using nonstd::visit;
    using nonstd::variant_npos;
    using nonstd::variant_size;
    using nonstd::variant_size_v;
    using nonstd::variant_alternative;
    using nonstd::variant_alternative_t;
}
```
# Problem

1. `monostate` conflicts:
Both `tl::optional` and variant-lite require a monostate type.
On MSVC 17.10.35122 (with C++17 enabled), the CRT also provides `std::monostate`.
This results in three different `monostate` definitions that are not interchangeable, causing compilation errors when the libraries are used together.

2. `bad_variant_access` mismatch:
The MSVC CRT provides `std::bad_variant_access`.
variant-lite uses its own `nonstd::bad_variant_access`.
Third-party code using our `std_compat::variant` must catch both exceptions to be safe, which is error-prone and unintuitive.

# Proposed solution:

Add configuration macros to variant-lite to allow users to override the `monostate` and `bad_variant_access` implementations with their own (or the standard library’s) versions.

For example:
```cpp
#ifdef variant_CONFIG_BAD_VARIANT_ACCESS
using variant_CONFIG_BAD_VARIANT_ACCESS;
#else
class variant_nodiscard bad_variant_access : public ::std::exception
{
public:
#if variant_CPP11_OR_GREATER
    virtual const char* what() const variant_noexcept variant_override
#else
    virtual const char* what() const throw()
#endif
    {
        return "bad variant access";
    }
};
#endif
```
```cpp
#ifdef variant_CONFIG_MONOSTATE
using variant_CONFIG_MONOSTATE;
#else
class monostate {};
#endif
```

# Benefits:

* Avoids duplicate/conflicting type definitions when integrating with other compatibility libraries.
* Allows seamless use of the standard library’s `monostate` and `bad_variant_access` when available.
* Reduces exception-handling boilerplate for downstream users.
* Keeps variant-lite flexible without breaking existing behavior.

If this approach does not conflict with any existing design decisions or constraints in **variant-lite**, I’m happy to prepare and submit the implementation as a PR.

MostafaNanticock committed:
Add support for configurable monostate type

Introduced variant_CONFIG_MONOSTATE that users can set its value if they want variant-lite to use it instead of its own monostate class

MostafaNanticock committed:
Add support for configurable bad_variant_access type

Introduced variant_CONFIG_BAD_VARIANT_ACCESS that users can set its value if they want variant-lite to use it instead of its own bad_variant_access class

martinmoene commented:
Note: continuation of [optional-lite issue 82](https://github.com/martinmoene/optional-lite/issues/82).

MostafaNanticock committed:
Update README.md synopsis: Configuration macros

Added `variant_CONFIG_BAD_VARIANT_ACCESS` and `variant_CONFIG_MONOSTATE` to the Synopsis: Configuration macros section

MostafaNanticock committed:
Merge branch 'master' into Add-the-ability-to-use-external-definitions-for-monostate-and-bad_variant_access

MostafaNanticock committed:
Updated template/variant.hpp

martinmoene commented:
My proofing sofar:

```C++
// main.cpp
// mix C++17 std::monostate and std::bad_variant_access with nonstd::variant with configurable such types.

#define variant_CONFIG_SELECT_VARIANT       variant_VARIANT_NONSTD 
#define variant_CONFIG_MONOSTATE            std::monostate
#define variant_CONFIG_BAD_VARIANT_ACCESS   std::bad_variant_access

#include <iostream>
#include <string>
#include <utility>      // std::monostate
#include <optional>
#include <variant>
#include "nonstd/variant.hpp"

namespace std_compat
{
    using std::monostate;
    using std::bad_variant_access;
}

namespace std_compat
{
    template <class T>
    using optional = std::optional<T>;

    using std::bad_optional_access;
    using std::in_place_t;
    using std::in_place;
    using std::nullopt;
    using std::nullopt_t;
    using std::make_optional;
    using std::swap;
}

namespace std_compat
{
    template <class... Types>
    using variant = nonstd::variant<Types...>;

    // using nonstd::bad_variant_access;
    // using nonstd::monostate;
    using nonstd::in_place_type_t;
    using nonstd::in_place_type;
    using nonstd::in_place_index_t;
    using nonstd::in_place_index;
    using nonstd::get;
    using nonstd::get_if;
    using nonstd::holds_alternative;
    using nonstd::visit;
    using nonstd::variant_npos;
    using nonstd::variant_size;
    using nonstd::variant_size_v;
    using nonstd::variant_alternative;
    using nonstd::variant_alternative_t;
}

int main()
{
    // using namespace std_compat;
    // use namespace std_compat explicitly, as to avoid 'invisible' parameter type-dependent (koenig) lookup

    auto opt = std_compat::optional<std::string>(std_compat::in_place, std::string("std::optional") );

    auto var = std_compat::variant<int, std::string>(std_compat::in_place_type<std::string>, std::string("std::optional") );

    try
    {
        auto result = std_compat::get<int>(var);
    } 
    catch( std_compat::bad_variant_access const & e ) 
    {
        std::cout << "std_compat::bad_variant_access: '" << e.what() << "'";
    }
}

// cl -nologo -EHsc -std:c++17 -I../../variant-lite/include main.cpp && main.exe
// g++ -std=c++17 -Wall -I../../variant-lite/include main.cpp -o main.exe && main.exe

#ifdef MSVC_OUTPUT
CMD> cl -nologo -EHsc -std:c++17 -I../../variant-lite/include -Dvariant_CONFIG_SELECT_VARIANT=variant_VARIANT_NONSTD main.cpp && main.exe
main.cpp
../../variant-lite/include\nonstd/variant.hpp(2723): error C2766: explicit specialization; 'std::hash<std::monostate>' has already been defined
C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.44.35207\include\variant(1706): note: see previous definition of 'std::hash<std::monostate>'
#endif

#ifdef GCC_OUTPUT
CMD> g++ -std=c++17 -Wall -I../../variant-lite/include main.cpp -o main.exe && main.exe
In file included from main.cpp:12:
../../variant-lite/include/nonstd/variant.hpp:2723:8: error: redefinition of 'struct std::hash<std::monostate>'
 2723 | struct hash< nonstd::monostate >
      |        ^~~~~~~~~~~~~~~~~~~~~~~~~
In file included from main.cpp:11:
c:\mingw\include\c++\11.2.0\variant:1829:12: note: previous definition of 'struct std::hash<std::monostate>'
 1829 |     struct hash<monostate>
      |            ^~~~~~~~~~~~~~~
main.cpp: In function 'int main()':
main.cpp:67:14: warning: unused variable 'result' [-Wunused-variable]
   67 |         auto result = std_compat::get<int>(var);
      |              ^~~~~~
#endif
```

Mostafa-Mahmoud95 commented:
I tried the code and it works fine on compiler explorer:
* [MSVC x86](https://godbolt.org/z/rf9MqazMv)
* [MSVC ARM64](https://godbolt.org/z/KaeErKrzE)
* [MSVC x64](https://godbolt.org/z/Gexobcaeo)
* [GCC x86_x64](https://godbolt.org/z/r8hT9nMa8)
* [GCC ARM](https://godbolt.org/z/vY99vznoE)
* [GCC ARMx64](https://godbolt.org/z/bY51j3hfr)

Never got any such errors from any compiler

MostafaNanticock committed:
Fixed: std::hash support with variant_CONFIG_MONOSTATE

MostafaNanticock commented:
@martinmoene 
don''t you think it would be a good idea to add this test case to the other unit tests?
```cpp
// main.cpp
// mix C++17 std::monostate and std::bad_variant_access with nonstd::variant with configurable such types.

#define variant_CONFIG_SELECT_VARIANT       variant_VARIANT_NONSTD 
#define variant_CONFIG_MONOSTATE            std::monostate
#define variant_CONFIG_BAD_VARIANT_ACCESS   std::bad_variant_access

#include <iostream>
#include <string>
#include <utility>      // std::monostate
#include <optional>
#include <variant>
#include "nonstd/variant.hpp"

namespace std_compat
{
    using std::monostate;
    using std::bad_variant_access;
}

namespace std_compat
{
    template <class T>
    using optional = std::optional<T>;

    using std::bad_optional_access;
    using std::in_place_t;
    using std::in_place;
    using std::nullopt;
    using std::nullopt_t;
    using std::make_optional;
    using std::swap;
}

namespace std_compat
{
    template <class... Types>
    using variant = nonstd::variant<Types...>;

    // using nonstd::bad_variant_access;
    // using nonstd::monostate;
    using nonstd::in_place_type_t;
    using nonstd::in_place_type;
    using nonstd::in_place_index_t;
    using nonstd::in_place_index;
    using nonstd::get;
    using nonstd::get_if;
    using nonstd::holds_alternative;
    using nonstd::visit;
    using nonstd::variant_npos;
    using nonstd::variant_size;
    using nonstd::variant_size_v;
    using nonstd::variant_alternative;
    using nonstd::variant_alternative_t;
}

int main()
{
    // using namespace std_compat;
    // use namespace std_compat explicitly, as to avoid 'invisible' parameter type-dependent (Koenig) lookup

    auto opt = std_compat::optional<std::string>(std_compat::in_place, std::string("std::optional") );

    auto var = std_compat::variant<int, std::string>(std_compat::in_place_type<std::string>, std::string("std::optional") );

    try
    {
        auto result = std_compat::get<int>(var);
    } 
    catch( std_compat::bad_variant_access const & e ) 
    {
        std::cout << "std_compat::bad_variant_access: '" << e.what() << "'";
    }
}
```
I can do that if you want

martinmoene commented:
@MostafaNanticock , It is indeed, thanks for mentioning it.

I'll do that later today.
